woocommerce/woocommerce-ios
# Statically-Typed Event Properties Convention and In-app Feedback Card Tracking

Ref #2548. 

## Changes

This adds two things.  

### Statically-Typed Event Properties Convention

I thought about incorporating @diegoreymendez' analytics tracking improvements (paNNhX-98-p2) into our codebase. Today seems to be a good day to implement it. Thanks to @diegoreymendez for the advice and for allowing me to _borrow_ the code. 🤪 

I believe Diego has explained in the post what sort of problems this solves so I won't explain that here. But let me know if you have any questions. 

A small difference in our implementation is I used `WooAnalyticsStat` instead of `String`:

https://github.com/woocommerce/woocommerce-ios/blob/2b2d41459d0eaeb551daf016f35e7e8144e0b9d6/WooCommerce/Classes/Analytics/WooAnalyticsEvent.swift#L40-L43

It is implemented like this in [WordPress-iOS-Shared](https://github.com/wordpress-mobile/WordPress-iOS-Shared/blob/fcfef1e440509d92514118d91631708ee67ebdaa/WordPressShared/Core/Analytics/AnalyticsEvent.swift#L41-L49):

```swift
public final class AnalyticsEvent {
    public let name: String
    public let properties: [String: String]
}
```

I wanted to keep the usage of `WooAnalyticsStat` for now because of Aaron's concern in paNNhX-98-p2#comment-283. I documented the rationale in `WooAnalyticsStat`:

https://github.com/woocommerce/woocommerce-ios/blob/2b2d41459d0eaeb551daf016f35e7e8144e0b9d6/WooCommerce/Classes/Analytics/WooAnalyticsStat.swift#L6-L9

### In-app Feedback Card Events

To test this convention, I added tracking for the In-app Feedback Card. 

<img src="https://user-images.githubusercontent.com/198826/88236893-4f12f280-cc3b-11ea-8308-ab4359e809f3.png" width="240">


Event Name | Trigger | Custom Properties
--------|-------|--
`app_feedback_prompt`  |  When the “Enjoying the WooCommerce app?” and the buttons are displayed to the user. | `action="shown"`
`app_feedback_prompt` | When the **I Like It** button is activated. | `action="liked"`
`app_feedback_prompt` | When the **Could Be Better** button is activated. | `action="didnt_like"`

## Testing

1. Delete the app from the device to reset any existing state. 
2. Run the app.
3. Navigate to Orders tab. 
4. Time travel to more than 3 months from now. 
5. Go back to the app and navigate to My Store. The In-app Feedback Card should be visible.
6. Confirm that you can see this log in the console:

    ```
    🔵 Tracked app_feedback_prompt, properties: [AnyHashable("action"): "shown", ...
    ```
7. Tap on the Could Be Better button. 
8. Confirm that you can see this log in the console:

    ```
    🔵 Tracked app_feedback_prompt, properties: [AnyHashable("action"): "didnt_like", ...
    ```
9. Navigate to Orders tab.
10. Time travel to 6 months from the current device's date. 
11. Go back to the app and navigate to My Store. The In-app Feedback Card should be visible again.
12. Tap on the I Like It button. 
13. Confirm that you can see this log in the console:
    ```
    🔵 Tracked app_feedback_prompt, properties: [AnyHashable("action"): "liked", ...
    ```
14. Don't forget to travel back to the present. We need you here! 🤪 

## Submitter Checklist
